synapse/handlers/message.py

```diff
@@ -1022,8 +1022,25 @@ async def create_new_client_event(
             #
             # TODO(faster_joins): figure out how this works, and make sure that the
             #   old state is complete.
-            old_state = await self.store.get_events_as_list(state_event_ids)
-            context = await self.state.compute_event_context(event, old_state=old_state)
+            metadata = await self.store.get_metadata_for_events(state_event_ids)
+
+            state_map = {}
+            for state_id in state_event_ids:
+                data = metadata.get(state_id)
+                if data is None:
+                    raise Exception("State event not persisted %s", state_id)
```

Comment:
It would be good to add some more comments, and/or improve the exception message here. What exactly does it mean? Somebody is trying to send an insertion event which refers to a state event which we haven't (yet?) seen? Is there a mechanism to pull said event which has failed, or is this a client-side error?

Reply:
Have added a comment to try and address those questions, hopefully it makes sense now?

https://github.com/matrix-org/synapse/pull/12852/commits/edcd824bdd4374f2656c7225ac94977664cc53dd#diff-3d4a81fcc5b115c649ae67aaba0dd4dd0c5827bb001dfcb90b00ffb1605d518eR1038-R1048